java/lang/StringBuffer/StringBuffer.cpp

```diff
@@ -0,0 +1,915 @@
+            && Character::isLowSurrogate(this->original[beginIndex])) {
+            codePointCount--;
+            beginIndex++;
+        }
+    }
+
+    return codePointCount;
+}
+
+StringBufferUnSafe &StringBufferUnSafe::deletes(int start, int end) {
+    if (start < 0) {
+        throw StringIndexOutOfBoundsException("index must be positive");
+    }
+
+    if (start > end || start > this->currentLength) {
+        throw StringIndexOutOfBoundsException();
+    }
+
+    if (start == end || start == this->currentLength) {
+        return *this;
+    }
+
+    string fromPosition;
+    int memorySizeToMove;
+
+    if (end > this->currentLength) {
+        fromPosition = this->original + this->currentLength;
+        memorySizeToMove = this->currentLength * sizeof(char);
+        this->currentLength = start;
+    } else {
+        fromPosition = this->original + end;
+        memorySizeToMove = (this->currentLength - start) * sizeof(char);
+        this->currentLength -= (end - start);
+    }
+
+    string toPosition = this->original + start;
+    memmove(toPosition, fromPosition, (size_t) memorySizeToMove);
+    return *this;
+}
+
+StringBufferUnSafe &StringBufferUnSafe::deleteCharAt(int index) {
+    if (index < 0) {
+        throw StringIndexOutOfBoundsException("index must be positive");
+    }
+
+    if (index >= this->currentLength) {
+        throw StringIndexOutOfBoundsException();
+    }
+
+    // move block of memory from index + 1 to index
+    string fromPosition = this->original + index + 1;
+    string toPosition = this->original + index;
+    int memorySizeToMove = (this->currentLength - index) * sizeof(char);
+    memmove(toPosition, fromPosition, (size_t) memorySizeToMove);
+    this->currentLength--;
+    return *this;
+}
+
+StringBufferUnSafe::~StringBufferUnSafe() {
+    free(this->original);
+}
+
+/*void StringBufferUnSafe::getChars(int sourceBegin, int sourceEnd, string destination, int destinationBegin) {
+    if (sourceBegin < 0 || destinationBegin < 0 || sourceBegin > sourceEnd
+        || (destinationBegin + sourceEnd - sourceEnd) > this->currentLength
+        || sourceEnd > this->currentLength ) {
+
+        throw IndexOutOfBoundsException();
+    }
+     // TODO need arrayCopy
+}*/
+
+int StringBufferUnSafe::indexOf(String stringToGetIndex) {
+    return string_index(this->original, stringToGetIndex.toString(), 1);
+}
+
+int StringBufferUnSafe::indexOf(String stringToGetIndex, int fromIndex) {
+    int stringAppearTime = 0;
+    int indexOfString = 0;
+    do {
+        stringAppearTime++;
+        indexOfString = string_index(this->original, stringToGetIndex.toString(), stringAppearTime);
+    } while (indexOfString < fromIndex && indexOfString != -1);
+
+    return indexOfString;
+}
+
+StringBufferUnSafe &StringBufferUnSafe::insert(int offset, float floatValue) {
+    int valueLength = String::valueOf(floatValue).length();
+    return this->insert(offset, String::valueOf(floatValue).toString(), 0, valueLength);
+}
```

Comment:
Should we create String object and call **.toString()** and **.length()** method?